File: services/amap_uri.py
"""高德 URI 导航链接构建"""
from typing import Any, Dict, Optional
from urllib.parse import quote

MODE_MAP = {
    "driving": "car",
    "car": "car",
    "walking": "walk",
    "walk": "walk",
    "transit": "bus",
    "bus": "bus",
    "riding": "ride",
    "ride": "ride",
}


def _point_segment(
    lnglat: Optional[list],
    location: str,
    name: str,
    fallback: str,
) -> Optional[str]:
    lng = lat = None
    if lnglat and len(lnglat) >= 2:
        lng, lat = lnglat[0], lnglat[1]
    elif location and "," in location:
        parts = location.split(",", 1)
        try:
            lng, lat = float(parts[0]), float(parts[1])
        except ValueError:
            return None
    if lng is None or lat is None:
        return None
    label = quote((name or fallback or "").strip() or fallback, safe="")
    return f"{lng},{lat},{label}"


def build_amap_navi_uri(
    origin: Dict[str, Any],
    destination: Dict[str, Any],
    mode: str = "driving",
) -> Optional[str]:
    """构建含起点/终点/出行方式的高德导航 URI。"""
    if not destination:
        return None
    to_seg = _point_segment(
        destination.get("lnglat"),
        destination.get("location", ""),
        destination.get("name", ""),
        "终点",
    )
    if not to_seg:
        return None
    m = MODE_MAP.get(mode, "car")
    parts = [
        f"to={to_seg}",
        f"mode={m}",
        "coordinate=gaode",
        "callnative=1",
        "src=AutoCityIntro",
    ]
    from_seg = _point_segment(
        origin.get("lnglat") if origin else None,
        (origin or {}).get("location", ""),
        (origin or {}).get("name", ""),
        "起点",
    )
    if from_seg:
        parts.insert(0, f"from={from_seg}")
    if m == "car":
        parts.append("policy=0")
    return f"https://uri.amap.com/navigation?{'&'.join(parts)}"


def attach_navi_uri(route_payload: Dict[str, Any]) -> Dict[str, Any]:
    if route_payload.get("type") == "route":
        uri = build_amap_navi_uri(
            route_payload.get("origin") or {},
            route_payload.get("destination") or {},
            route_payload.get("mode", "driving"),
        )
        if uri:
            route_payload["amap_navi_uri"] = uri
    return route_payload


def build_taxi_uri(lng: float, lat: float, name: str = "目的地") -> str:
    label = quote((name or "目的地").strip(), safe="")
    return f"https://uri.amap.com/taxi?dlat={lat}&dlon={lng}&dname={label}&dev=0&callnative=1"

File: static/amap-uri.js
/** 高德 URI 唤起（Web / App 导航） */
const AmapUri = (() => {
  const MODE_MAP = {
    driving: "car",
    car: "car",
    walking: "walk",
    walk: "walk",
    transit: "bus",
    bus: "bus",
    riding: "ride",
    ride: "ride",
  };

  function pointSegment(lnglat, locationStr, name, fallbackName) {
    let lng;
    let lat;
    if (Array.isArray(lnglat) && lnglat.length >= 2) {
      lng = lnglat[0];
      lat = lnglat[1];
    } else if (locationStr && String(locationStr).includes(",")) {
      const parts = String(locationStr).split(",");
      lng = parseFloat(parts[0]);
      lat = parseFloat(parts[1]);
    }
    if (lng == null || lat == null || Number.isNaN(lng) || Number.isNaN(lat)) return null;
    const label = encodeURIComponent((name || fallbackName || "").trim() || fallbackName);
    return `${lng},${lat},${label}`;
  }

  /**
   * 构建导航 URI：含起点、终点、出行方式，并尝试唤起高德 App。
   * @param {object} routeData - route_map / parse_route_result 结构
   */
  function buildNavigation(routeData) {
    if (!routeData?.destination) return null;
    return buildNavigationUri(
      routeData.origin || {},
      routeData.destination,
      routeData.mode || routeData.mode_label || "driving"
    );
  }

  /**
   * 根据起终点与出行方式构建高德导航 URI。
   * @param {object} origin - { lnglat?, location?, name? }
   * @param {object} destination - { lnglat?, location?, name? }
   * @param {string} mode - driving/walking/transit/...
   */
  function buildNavigationUri(origin, destination, mode = "driving") {
    if (!destination) return null;
    const navMode = MODE_MAP[mode] || "car";
    const toSeg = pointSegment(
      destination.lnglat,
      destination.location,
      destination.name,
      "终点"
    );
    if (!toSeg) return null;

    const params = [
      `to=${toSeg}`,
      `mode=${navMode}`,
      "coordinate=gaode",
      "callnative=1",
      "src=AutoCityIntro",
    ];

    const fromSeg = pointSegment(
      origin?.lnglat,
      origin?.location,
      origin?.name,
      "起点"
    );
    if (fromSeg) params.unshift(`from=${fromSeg}`);

    if (navMode === "car") params.push("policy=0");
    if (navMode === "bus") params.push("policy=0");

    return `https://uri.amap.com/navigation?${params.join("&")}`;
  }

  function openInAmap(uri) {
    if (!uri) return;
    const link = document.createElement("a");
    link.href = uri;
    link.target = "_blank";
    link.rel = "noopener noreferrer";
    document.body.appendChild(link);
    link.click();
    link.remove();
  }

  return { buildNavigation, buildNavigationUri, openInAmap, MODE_MAP };
})();
